src/common/async/waiter.h

```diff
@@ -69,12 +69,12 @@ class rvalue_reference_wrapper {
   }
 
   template<typename... Args>
-  std::result_of_t<T&(Args&&...)> operator ()(Args&&... args ) const {
+  std::invoke_result_t<T&, Args...> operator()(Args... args) const {
```

Comment:
the change from `Args&&...` to `Args...` looks like it would break the use of `std::forward<Args>(args)...` below. cc @adamemerson

Reply:
Hrm, I think this was prompted by my comment: https://github.com/ceph/ceph/pull/51197#discussion_r1209510635. I only meant `std::invoke_result_t` specification (i.e. stuff between angle brackets) there, I didn't mean it to extend to parameters.

It would be great if Adam could take a closer look just in case.

Reply:
Whops, I've just replaced this with:

```
std::invoke_result_t<T, Args...> operator()(Args&&... args) {
```

Reply:
Sounds good, then!

Reply:
I'm not completely clear.  Here is before and after, adjusted for formatting fixups:
```
std::result_of_t<T&(Args&&...)> operator()(Args&&... args) const {
std::result_of_t<T&&(Args&&...)> operator()(Args&&... args) {

std::invoke_result_t<T&, Args...> operator()(Args&&... args) const {
std::invoke_result_t<T, Args...> operator()(Args&&... args) {
```
So `T&(Args&&...)` became `T&, Args...`.  However `T&&(Args&&...)` became `T, Args...`, not `T&&, Args...`.  Is that intended?  Apologies for being dense, but one can't be too careful when it comes to a type named `rvalue_reference_wrapper` ;)

Going further, I wonder why we are bothering with `std::invoke_result_t` here at all?  Can't this overload set simply use `auto` (still preserving the desired lvalue/rvalue semantics)?
```
auto operator()(Args&&... args) const {
auto operator()(Args&&... args) {
```

Reply:
``auto`` seems to work fine now, thanks!